joomla/joomla-cms
# [5.2] SEF: Add option to don't set Itemid to homepage by default

This fixes issues #43154, #40991, #43104, #35570, #30818, #42686, #41897.

### Summary of Changes
Somewhere before Joomla 2.5.0 a person decided that we need to force an Itemid in each and every case of building a URL, defaulting to the home menu item if none is set. This currently results in the URLs for core components never being allowed to hit the NoMenu rule. The end result is, that these URLs are sometimes not SEFed at all.


This PR is depending on the switch in #43432. (has been merged)

This PR contains a few changed system tests. These tests were either broken, tested broken behavior or with the latest changes were unnecessary.

I'd like to thank [Joomla Agentur](https://joomla-agentur.de/) for sponsoring this change.


### Testing Instructions
1. Install the testing sampledata
2. create an article in the uncategorised category
3. search for that article in smart search. Notice that the URL is not SEFed.
4. switch the "Disallow non-SEF URLs" in the SEF system plugin to yes
5. Reload the search result page. Notice that the URL now is at least partially SEFed with the URL starting with /component/content/article/[alias]?catid=someID


By using the switch, this is backwards compatible the way it is now. When Joomla 6.0 comes around, the code should be removed completely and the changed behavior would then be the new default.


### Link to documentations
Please select:


I checked the "no documentation changes needed" for now. I understand that some documentation for this needs to happen, but I don't want to add this to the live documentation yet when the feature has not been accepted yet.
